File: ToC_2026/8/question4/C/1.c
//階乗関数 fac(m) = m!
#include <stdio.h>

int iszero(int m) { return m == 0; }
int pi(int m) { if (m == 0) return 0; return m - 1; }

int fac(int m) {
    if (iszero(m)) {
        return 1;
    } else {
        return m * fac(pi(m));
    }
}

int main() {
    int m;
    printf("Enter a number: ");
    scanf("%d", &m);
    printf("The factorial of %d is %d\n", m, fac(m));
    return 0;
}

File: ToC_2026/8/question4/Rust/1.rs
use std::io;

fn iszero(m: i32) -> bool { m == 0 }
fn pi(m: i32) -> i32 { if m == 0 { 0 } else { m - 1 } }

fn fac(m: i32) -> i32 {
    if iszero(m) {
        1
    } else {
        m * fac(pi(m))
    }
}

fn main() {
    println!("Enter a number: ");
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let m: i32 = input.trim().parse().unwrap();
    
    println!("The factorial of {} is {}", m, fac(m));
}